File: src/main/python/AoC2020_12.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 12
#

from __future__ import annotations
from typing import NamedTuple
import aocd
from aoc import my_aocd
from aoc.common import log
from aoc.geometry import Position, Turn
from aoc.navigation import (
    Heading,
    Waypoint,
    NavigationWithHeading,
    NavigationWithWaypoint,
)
from enum import Enum


class Action(Enum):
    NORTH = "N"
    EAST = "E"
    SOUTH = "S"
    WEST = "W"
    RIGHT = "R"
    LEFT = "L"
    FORWARD = "F"

    @classmethod
    def from_str(cls, s: str) -> Action:
        for v in Action:
            if v.value == s:
                return v
        raise ValueError


class NavigationInstruction(NamedTuple):
    action: Action
    value: int


def _parse(inputs: tuple[str, ...]) -> list[NavigationInstruction]:
    return [
        NavigationInstruction(Action.from_str(i[0]), int(i[1:]))
        for i in inputs
    ]


def _navigate_with_heading(
    navigation: NavigationWithHeading, nav: NavigationInstruction
) -> None:
    if nav.action == Action.RIGHT:
        navigation.turn(Turn.from_degrees(nav.value))
    elif nav.action == Action.LEFT:
        navigation.turn(Turn.from_degrees(360 - nav.value))
    elif nav.action == Action.FORWARD:
        navigation.forward(nav.value)
    elif nav.action in {Action.NORTH, Action.EAST, Action.SOUTH, Action.WEST}:
        navigation.drift(heading=Heading[nav.action.name], amount=nav.value)


def part_1(inputs: tuple[str, ...]) -> int:
    navs = _parse(inputs)
    navigation = NavigationWithHeading(
        position=Position(0, 0), heading=Heading.EAST
    )
    log(navigation)
    for nav in navs:
        _navigate_with_heading(navigation, nav)
        log(navigation)
    return abs(navigation.position.x) + abs(navigation.position.y)


def _navigate_with_waypoint(
    navigation: NavigationWithWaypoint, nav: NavigationInstruction
) -> None:
    if nav.action == Action.RIGHT:
        navigation.right(nav.value)
    elif nav.action == Action.LEFT:
        navigation.left(nav.value)
    elif nav.action == Action.FORWARD:
        navigation.forward(nav.value)
    elif nav.action in {Action.NORTH, Action.EAST, Action.SOUTH, Action.WEST}:
        navigation.update_waypoint(
            heading=Heading[nav.action.name], amount=nav.value
        )
    else:
        raise ValueError("invalid input")


def part_2(inputs: tuple[str, ...]) -> int:
    navs = _parse(inputs)
    start = Waypoint(10, 1)
    navigation = NavigationWithWaypoint(Position(0, 0), start)
    log(navigation)
    for nav in navs:
        _navigate_with_waypoint(navigation, nav)
        log(navigation)
    return abs(navigation.position.x) + abs(navigation.position.y)


TEST = """\
F10
N3
F7
R90
F11
""".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2020, 12)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST) == 25  # type:ignore[arg-type]
    assert part_2(TEST) == 286  # type:ignore[arg-type]

    inputs = my_aocd.get_input_data(puzzle, 785)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2020_12.java
import static java.util.stream.Collectors.toList;

import java.util.Arrays;
import java.util.List;

import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.geometry.Turn;
import com.github.pareronia.aoc.navigation.Heading;
import com.github.pareronia.aoc.navigation.NavigationWithHeading;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2020_12 extends AoCBase {
	
	private final List<NavigationInstruction> instructions;
	
	private AoC2020_12(final List<String> input, final boolean debug) {
		super(debug);
		this.instructions = input.stream()
		        .map(s -> NavigationInstruction.of(s.substring(0, 1), s.substring(1)))
		        .collect(toList());
	}
	
	public static AoC2020_12 create(final List<String> input) {
		return new AoC2020_12(input, false);
	}

	public static AoC2020_12 createDebug(final List<String> input) {
		return new AoC2020_12(input, true);
	}
	
	@Override
	public Integer solvePart1() {
	    return new Navigation(
                this.instructions, Heading.EAST, ActionMeaning.ONE)
            .getDistanceTraveled();
	}
	
	@Override
	public Integer solvePart2() {
	    return new Navigation(
                this.instructions, Heading.of(10, 1), ActionMeaning.TWO)
            .getDistanceTraveled();
	}

	public static void main(final String[] args) throws Exception {
		assert AoC2020_12.createDebug(TEST).solvePart1() == 25;
		assert AoC2020_12.createDebug(TEST).solvePart2() == 286;

        final Puzzle puzzle = Aocd.puzzle(2020, 12);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2020_12.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2020_12.create(inputData)::solvePart2)
        );
	}
	
	private static final List<String> TEST = splitLines(
	        "F10\r\n" +
	        "N3\r\n" +
	        "F7\r\n" +
	        "R90\r\n" +
	        "F11"
	);
	
	public enum Action {
	    NORTH("N"), EAST("E"), SOUTH("S"), WEST("W"),
	    RIGHT("R"), LEFT("L"), FORWARD("F");
	    
	    private final String code;
	    
	    Action(final String code) {
	        this.code = code;
	    }
	    
	    public static Action of(final String code) {
	        return Arrays.stream(Action.values())
	                .filter(a -> a.code.equals(code))
	                .findFirst()
	                .orElseThrow();
	    }
	}
	
	record NavigationInstruction(Action action, int value) {
	    
	    public static NavigationInstruction of(final String action, final String value) {
	        return new NavigationInstruction(
	            Action.of(action), Integer.parseInt(value)
	        );
	    }
	}
	
	private enum ActionMeaning {
	    ONE, TWO;
	}
	
	private static final class Navigation {
	    private final List<NavigationInstruction> instructions;
	    private final ActionMeaning actionMeaning;
	    private final NavigationWithHeading nav;
	    private final Position start = Position.ORIGIN;
	    
	    public Navigation(
	            final List<NavigationInstruction> instructions,
	            final Heading initialHeading,
                final ActionMeaning actionMeaning
        ) {
            this.instructions = instructions;
            this.actionMeaning = actionMeaning;
            this.nav = new NavigationWithHeading(this.start, initialHeading);
        }

        public int getDistanceTraveled() {
            this.instructions.forEach(this::executeInstruction);
            return this.nav.getPosition().manhattanDistance(this.start);
	    }

        private void executeInstruction(final NavigationInstruction instuction) {
            final Action action = instuction.action();
            final int value = instuction.value();
            if (action == Action.RIGHT) {
                this.nav.turn(Turn.fromDegrees(value));
            } else if (action == Action.LEFT) {
                this.nav.turn(Turn.fromDegrees(360 - value));
            } else if (action == Action.FORWARD) {
                this.nav.forward(value);
            } else if (this.actionMeaning == ActionMeaning.ONE) {
                this.nav.drift(Heading.fromString(action.code), value);
            } else {
                final Direction dir = Direction.fromString(action.code);
                this.nav.setHeading(nav.getHeading().add(dir, value));
            }
        }
    }
}